[MQTT] (Auto)Discovery implementation

tonhuisman opened the pull request:
Resolves #977 
Resolves #4982 
And Forum request(s) ([here](https://www.letscontrolit.com/forum/viewtopic.php?t=9843))

Features:
- MQTT (Auto)Discovery for all supported devices configured to use a MQTT Controller that has MQTT Discovery implemented
  - Controller settings for Trigger and Discovery topics, and an Enabled checkbox
    - Suggested defaults:
    - Discovery Trigger topic: `homeassistant/status` (listening for payload `online`).
    - Auto Discovery topic: `homeassistant/%devclass%/%unique_id%` (`/config` will be added by default) `%unique_id%` is a concatenation of the sysname, taskname and valuename, separated by an underscore `_`, and converted to lowercase.
    - Config suffix|online message (/config|online): `/config` (will use /config and online by default when left empty)
    - Retain Discovery: Unchecked, can be checked to send the AutoDiscovery with Retain flag.
    - Actions:
    - When discovery is triggered, a random delay between 1 and 10 seconds is inserted before starting the action, to spread the load on the receiving server.
    - Discovery is sent for all enabled tasks that have this controller enabled in their settings.
    - When the controller is enabled and Auto Discovery is enabled, on clicking the Submit button a full discovery is sent.
    - The Discovery is triggered when the Controller is started.
    - Discovery is sent using the Retain Discovery setting.
  - Device settings for Controller config to set an optional Group ID and Resend MQTT Discovery
    - When the task is enabled, the Resend MQTT Discovery checkbox is checked and the Submit button used, the Discovery for the current device is resent.
    - The Group field can be used to manually group multiple tasks as a single device on HA. When 0, all values of the device are grouped as a single, unique, device.
- [C005] Home Assistant (openHAB) MQTT: (partially) implemented (more device classes to add).
  - Device classes: switch, temperature, humidity, pressure, distance, pm1, pm25, pm10
- [P052] CO2 Senseair: implement device class config for supported values
- [P001] Switch: implemented all HA-supported MQTT Device Classes for `binary_sensor`, and Inversed Logic setting
- Plugins with user-configurable or undefined Value Types now have a selector per Value that is used by AutoDiscovery to determine how to send the discovery message
- [Derived Values] Add support for Value Type, so they can be used in MQTT AutoDiscovery
- [AutoDiscovery] Add support for Derived values, Unit of Measure and Value Type (Home assistant)
- Sending Derived values can be enabled per controller, if the controller supports that
- [C005] Add support for sending Derived values
- [C006] Add support for sending Derived values
- [C014] Add Unit of Measure to Homie protocol AutoDiscovery (`/$unit`)
- [C014] Add support for sending Derived values

TODO:
- [ ] Testing
- [x] Enable actually sending the MQTT messages...
- [x] Documentation updates
  - [x] Add section to Development documentation (added to _Pxxx_PluginTemplate source)
  - [x] Update C005 documentation
  - [x] Update DevicePage documentation for Value Type selection
- [x] Implement more `device class` handlers: lux, moisture, co2, weight, uv, uvindex, ir, distance, dust, aqi, nox, voltage, current, power, color, color temperature, (wind)direction, wind speed.
- [ ] Additional `sensor` handlers: dimmer, gps, analog. (may be implemented later)
- [x] Implement more plugin specific value handlers (for plugins that allow user-selectable values)
- [x] Add configurable trigger value (default: `online`)
- [x] Add configurable Config topic suffix (`/config`)
- [x] Add MQTT Device Class setting and Inverted to Sensor_VType::Switch devices

tonhuisman committed:
[MQTT] (Auto)Discovery implementation

tonhuisman committed:
[P052] Add MQTT Discovery data for supported Values

tonhuisman committed:
[Source] Fix CR/LF vs LF

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
[MQTT] Add support for more VTypes, implement more VTypes, small improvements

TD-er commented on src/src/Helpers/_CPlugin_Helper_mqtt.cpp:
```diff
@@ -591,7 +738,12 @@ bool MQTT_DiscoveryGetDeviceVType(taskIndex_t                 TaskIndex,
       discoveryItems.push_back(DiscoveryItem(Sensor_VType::SENSOR_TYPE_DUSTPM2_5_ONLY, valueCount, 0));
       break;
-    case 47:  // Soil Moisture
+    case 27:  // INA219
```
Not sure if it is the correct place to do these plugin specifics here.
Maybe best to 'ask' the plugin (or task) to format those. A bit like the raw option as used for LoRaTTN.

This way the knowledge remains inside the plugin code and you don't het those ugly `#ifdef USES_Pxxx` checks for each case in this switch statement.

tonhuisman replied:
At first I was quite reluctant to change (nearly) all plugins, but it's indeed the ~best~ right way to go forward. (And I practiced on changing all those files, recently 😅)
I've already added a plugin function `PLUGIN_GET_DISCOVERY_VTYPES` to retrieve that info (see P052 for a custom/user-defined implementation, for standard plugins it's much simpler)

I've 'configured' most of the standard plugins in a single place for now, making that process easier. I'll move that to the separate files 😉 👍

TD-er replied:
Maybe this can later also be used to implement the feature to use UnitOfMeasure.
This can then be used to offer default sensible settings for the stats/chart layout, unit conversion (C-to-F for example) and some standard options to globally set preferred units.

So internally we will use standard Si units and when using the value it will be converted to the requested unit of measure.

tonhuisman replied:
> this can later also be used to implement the feature to use UnitOfMeasure.

That's one of the things that keeps popping up in the back of my head while assembling those Json discovery excerpts 😃

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
[MQTT] Move TaskValue VType Discovery to each plugin where needed. Some other code improvements

TD-er commented on src/_P018_Dust.ino:
```diff
@@ -58,6 +59,17 @@ boolean Plugin_018(uint8_t function, struct EventStruct *event, String& string)
+    case PLUGIN_GET_DISCOVERY_VTYPES:
+    {
+      for (uint8_t i = 0; i < event->Par5; ++i) {
+        event->ParN[i] = static_cast<int>(Sensor_VType::SENSOR_TYPE_DUSTPM2_5_ONLY);
```
Typo `ParN` ?

tonhuisman replied:
New 'feature', it's a `union` with `Par1`..`Par5`, to be able to set these in a for loop.
[Here](https://github.com/letscontrolit/ESPEasy/pull/5199/files#diff-4e703efd77a3caeffd2aaf98b6a435776cd26ab0ecb0ffe76ed96db8f5a3e203R63-R72)

uzi18 replied:
@tonhuisman maybe same for strings?

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
[P105] Fix Uncrustify formatting

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
[P164] Add fixme for currently unsupported AQI Discovery value

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
[Discovery] Add VType for Reactive Power

tonhuisman committed:
[Discovery] Implement VType for Reactive Power

tonhuisman committed:
[P076] Implement Discovery support (for user-selected values)

tonhuisman committed:
[P077] Implement Discovery support (for user-selected values)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
[P078] Implement Discovery support (for user-selected values)

tonhuisman committed:
[P085] Implement Discovery support (for user-selected values)

tonhuisman committed:
[P102] Implement Discovery support (for user-selected values)

tonhuisman committed:
[P108] Implement Discovery support (for user-selected values)

tonhuisman committed:
[P115] Implement Discovery support (for user-selected values)

tonhuisman committed:
[P132] Implement Discovery support (for user-selected values)

tonhuisman committed:
[P142] Implement Discovery support (for user-selected values)

TD-er commented on src/_P077_CSE7766.ino:
```diff
@@ -84,11 +86,21 @@ boolean Plugin_077(uint8_t function, struct EventStruct *event, String& string)
-      event->Par1 = static_cast<int>(Sensor_VType::SENSOR_TYPE_NONE); // Not yet supported
-      success     = true;
+      for (uint8_t i = 0; i < event->Par5; ++i) {
+        const uint8_t choice = PCONFIG(i + P077_QUERY1_CONFIG_POS);
```
Maybe also add some check for i not getting out of bounds.
```c++
for (uint8_t i = 0; i < event->Par5 && (i < (8 - P077_QUERY1_CONFIG_POS)); ++i) 
```

And this for multiple other similar calls

tonhuisman replied:
`Par5` is set using a call with function `PLUGIN_GET_DEVICEVALUECOUNT` (by calling function `getValueCountForTask()`), so should not get out of range.

tonhuisman committed:
[Discovery] Refactor getting VType per plugin, add some more plugins for AutoDiscovery

TD-er commented on src/src/Helpers/_CPlugin_Helper_mqtt.cpp:
```diff
@@ -242,6 +242,51 @@ bool MQTT_protocol_send(EventStruct *event,
+  constexpr uint8_t _max = PLUGIN_CONFIGVAR_MAX;
+
+  for (uint8_t i = 0; i < nrVars && i < _max; ++i) {
+    const uint8_t choice = (255 == pConfigOffset) ? i : PCONFIG(i + pConfigOffset);
```
Why not simply using pConfigOffset of `0` ?

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
[Discovery] Add support for some more VTypes, add AQI and NOx

tonhuisman committed:
[P147] Implement MQTT Discovery support

tonhuisman committed:
[P164] Update MQTT Discovery support

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'feature/MQTT-discovery' of https://github.com/tonhuisman/ESPEasy-1 into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' into feature/MQTT-discovery

tonhuisman commented:
@jelinekpetr Are you available for testing this PR? Would be much appreciated 💯

jelinekpetr commented:
Hi, my approval was a fat finger action in VS Code, but actually I'm testing this feature just now and yes, I would like to help to speed up the development. 
I've just added temp/humidity sensors (SI70xx) with only one small difference comparing to recommendation for controller setting (I had to change the "Auto discovery topic" to homeassistant/%devclass%/%sysname%/%valname% to correct discovery from HA - I use Zigbee2MQTT integration). Now I'm trying to connect external contact connected to GPIO. Everything looks OK but HA cannot want to add a new device which is announced in the discovery topic... I'll try to find where do I a mistake.
I have a lot of ESP8266 based sensors connected to Domoticz but I'm thinking about migration - not much hurry about that but very likely I will do a migration because my A/C is possible to control from HA not from Domoticz and two automation systems in one house could be a bit headache...

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

jelinekpetr commented:
Hi, I'm not sure whether this place is the right one to report my findings but I'm writing it here at least for now.
I've found that Input switch is not recognized by HA because of:
1. Wrong device class declaration. There is no "switch" device class of binary sensor at least for MQTT. You have to select one of them mentioned here: https://www.home-assistant.io/integrations/binary_sensor/#device-class. Or get rid of "dev_cla" completely which causing in HA showing default icon: 
![obrazek](https://github.com/user-attachments/assets/3c5d81e8-d69d-4a0c-9c75-a56aa2839308)
The best solution would be enabling user to select type of switch (device class) in the ESPEasy device configuration GUI. But I suppose this change would be a bit complicated. Moreover even the default class makes the switch accessible for automation and scripts.
2. In case of Input switch it is necessary to add two attributes: "payload_on" and "payload_off". I think it would be a constant string something like this (which works OK):
`const String payload          = F("payload_on\":\"1\",\"payload_off\":\"0");`

I have verified both changes using some dodgy changes in _CPlugin_Helper_mqtt.cpp and it works fine.

And what about sending the discovery messages retained? Just an idea... I can see this type of messages coming from Zigbee2MQTT as retained.

Thanks.

Petr

TD-er commented:
@jelinekpetr Good feedback and this is a perfectly fine place to give such feedback.

thomastech commented:
Just saw this commit. Very cool.

I'm currently an OpenHabian user, but looking forward to migrating to Home Assistant. Not much progress yet; months ago I purchased the hardware (Wyse 5070 Thin Client 8GB with SSD) and loaded the basic software. Will probably wait until the end of the year to move forward.

Regardless, I am happy to see you are working on this controller feature. Because no doubt I will be using it. :)

- Thomas

tonhuisman commented:
> (I had to change the "Auto discovery topic" to homeassistant/%devclass%/%sysname%/%valname% to correct discovery from HA

How I've interpreted the discovery, is that the `%sysname%/%valname%` part should match the Controller Publish setting (thought about taking that setting, but chose for max flexibility instead), so HA could match that when receiving the data, but maybe my expectations are too high here 😅 🤔 
But maybe you've adjusted your Controller Publish setting?



> 1. Wrong device class declaration. There is no "switch" device class of binary sensor at least for MQTT.

I've tried to interpret that documentation, but it's not always as clear as one might want it to be...



> 2\. In case of Input switch it is necessary to add two attributes: "payload_on" and "payload_off".

I hadn't found that in the documentation yet, thanks for this helpful info!
For a Switch device there's an Inversed Logic checkbox that has to be considered here, so I'll make the needed adjustments 👍

tonhuisman commented:
> The best solution would be enabling user to select type of switch (device class) in the ESPEasy device configuration GUI.

That's a change I intend to start working on, as a separate project, once this Discovery is finished and merged into ESPEasy

tonhuisman committed:
[Discovery] Add /config and Retain Discovery settings to MQTT Controller with AutoDiscovery

tonhuisman committed:
[Discovery] Fix Switch auto-discovery
[Discovery] Add support for Inverted Switch
[Discovery] Add support for Switch Device Class configuration

tonhuisman committed:
[P001] Implement Switch MQTT Device Class setting and Inverted Switch discovery

tonhuisman commented:
@jelinekpetr I've fixed a lot of issues I didn't even know where there... 😅, like adding a trigger for Switch devices, and correcting the Discovery topic, by adding support to use the Unique ID. (And I finally got it working on my rather new HA install ☮️)

Also added configuration for Retain Discovery to the Controller, and MQTT Device Class configuration to the Switch plugin (P001), that's to be added to other Switch-like plugins too.

Can you please update and test these updates?

tonhuisman committed:
[P009] Implement Switch MQTT Device Class setting and Inverted Switch discovery

tonhuisman committed:
[P001][P009] Correct typo

tonhuisman committed:
[P019] Implement Switch MQTT Device Class setting and Inverted Switch discovery

tonhuisman committed:
[P021] Implement Switch MQTT Device Class setting and Inverted Switch discovery

tonhuisman committed:
[Discovery] Add support for different Switch device class per value

jelinekpetr commented:
Sorry to be silent so long time - have lots of another work to finish. I'm going to start testing... Will report ASAP.

tonhuisman committed:
[Discovery] Fix typo in BinarySensorInverted helper

tonhuisman committed:
[Discovery] Implement UoM per taskValue when available

tonhuisman committed:
[Discovery] Add default empty option for Device Class selector (all values shift 1 index up!)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
[VTypes] Refactor SensorValueType selector so it can be used separately (cherry picked)

tonhuisman committed:
[ValueTypes] Add Custom VType selection

tonhuisman committed:
[Plugins] Add CustomVTypeVar option for plugins that need custom Value Type selection

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
[DerivedValues] Add support for Value Type argument

tonhuisman committed:
[AutoDiscovery] Add support for Derived values, UoM and Value Type, some code optimizations

tonhuisman committed:
[C006] Add support for sending derived values

tonhuisman committed:
[C014] Add Unit of Measure to AutoDiscovery, add Send Derived values support

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
[Docs] Update docs for Value Type setting for selected plugins

tonhuisman committed:
[Docs] Move Send derived values as a per controller setting for tasks

tonhuisman committed:
[Derived Values] Move enable sending per task-controller, Code optimization

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/MQTT-discovery

tonhuisman committed:
[AutoDiscovery] Add support for Wind direction and Wind speed

tonhuisman committed:
[Plugins] Minor code improvement

tonhuisman committed:
[UoM] Restructure UoM selector into categories, add more uom options

tonhuisman committed:
[UoM] Fix UoM selector empty value

tonhuisman committed:
[Docs] Update Unit of Measure documentation and screenshots

tonhuisman committed:
[AutoDiscovery] Embed setting for `online` message in `/config` setting

tonhuisman committed:
[Docs] Update documentation for C005

tonhuisman committed:
[AutoDiscovery] Update PluginTemplate

tonhuisman committed:
[AutoDiscovery] Embed setting for `online` message in `/config` setting

tonhuisman committed:
[AutoDiscovery] Remove unused code

tonhuisman committed:
[AutoDiscovery] Make start delay compiletime configurable, and extend max delay from 10 to 30 sec.

tonhuisman committed:
[Docs] Update documentation for C005

KnatteAnka commented:
Some question to compare to my Node red Autodiscovery "Addon"

1. Does it handle setting values of a Dummy device?
2. Can you hide values as internal values per devide or even per variable? ( i use "_" to tell its internal for autodiscovery)
3. Does it check so mqtt string and values do not contain any character other than A-Z,_,0-9?
4. can you define a dummy devide as a Dimmer? 
5. can you define a Select to input to espeasy different states?

TD-er commented:
Not sure if Ton is already back home to reply...
So the only thing I know about this is about point 3, where I found out the hard way that a space should not be used.
Meaning Ton will change how a hostname may be used in these topics by replacing the space.  Not sure what he will use, either '_' or '-'.
Would it be better to make this configurable? Or is a minus also not allowed?


Ton has added both unit-of-measure selections and value type. So I guess it should be possible to define a dummy task as dimmer. Not sure though what HA then expects how such a device should react on dimming updated from HA, or send dimming updates back to HA. The latter may not be an insignificant issue as I know from Domoticz that it was quite hard to make sure a device would not start to 'oscillate' by constantly toggling a relay for example.

Anyway, since this PR already has been merged, maybe best to open a new issue so we can have all information in 1 place.

KnatteAnka commented:
hmm '-' should be allowed yes https://www.home-assistant.io/integrations/mqtt/

and not sure i will update and test this  addon anytime soon :(  so no worry if answers take time, when/if i get time to test around and see any problems i will create a new issue then.

right checked my documentation at:
https://github.com/KnatteAnka/NodeRED_ESPEASY_AutoDiscovery/blob/main/README.md

and for dimmers i used 2 devices on EE one MQTT import and one dummy device to tell HA what state it is right now, this as i could not get EE to understand messages from HA other than MQTT import or with run event command on device

for Buttons and switches  it triggers a event on the EE with the name of the variable

tonhuisman commented:
> Some question to compare to my Node red Autodiscovery "Addon"
> 
>     1. Does it handle setting values of a Dummy device?

Changes to the MQTT Controllers in the recent past include handling of the `/cmd` and `/set` endpoints for published task/values for performing general commands, and/or setting a value to the task/value, assuming it allows to be set (most sensors overwrite these values periodically per Interval setting).

>     2. Can you hide values as internal values per devide or even per variable? ( i use "_" to tell its internal for autodiscovery)

For predefined Value types you can't, currently, but for devices that allow setting the Value type (per value), you can set the Value Type to `None` so it won't be used in Auto Discovery. This does _not_ stop the value from being published, though, currently there is no mechanism to do that. Not sure if that would be needed.
OTOH, HA allows to hide discovered entities, so that would also be a quite feasible/common way to handle that.

>     3. Does it check so mqtt string and values do not contain any character other than A-Z,_,0-9?

No, but I'll add that check in #5349 soon.

>     4. can you define a dummy devide as a Dimmer?

A dimmer is something that's rather hard to realize in ESPEasy right now, as we don't have a button that supports auto-repeat when being pressed. As mentioned in my reply to 1., you can send commands, and/or set value(s) to a task/value.

>     5. can you define a Select to input to espeasy different states?

As long as the _value_ is numeric, that will work fine (the core of ESPEasy is using numeric/float values).
Using the new Presentation feature, added recently when introducing String-variables (see `TaskValueSetPresentation` command), you can map that value to a string to present to the user on the Devices page. Or in [EasyFetch](https://github.com/chromoxdor/EasyFetch), but I assume you're not currently using that 😉. 


Thanks for the pointers given (again) 👍

tonhuisman commented:
> hmm '-' should be allowed yes

Well, ESPEasy doesn't like the dash as part of a system, task or value name, it will in many situations be interpreted as a subtract operation 😮

repa6 commented:
@tonhuisman, HA MQTT discovery was changed.
I can get my fake SDM230 meter (TAC2100) working as follows:

config topic:
`homeassistant/sensor/Energy_11/SDM230_ImpA/config`
payload:
`{	
	"dev":{
		"name":"Energy_11 SDM230",
		"ids":["Energy_11_SDM230"],
		"mf":"ESPEasy",
		"sw":"20250725"		
	},	
			"name":"SDM230 Import Active Energy",
			"unique_id":"energy_11_sdm230_ImpA",
			"device_class": "energy",
			"state_class": "total_increasing",
			"unit_of_measurement":"kWh",
			"state_topic":"EspEasy/Energy_11/SDM230s3/ImpA"
}`

tonhuisman commented:
> `"state_class": "total_increasing",`

Hmm, we currently don't have a setting for that, so I'll have to add some more code related to MQTT Discovery 😅

repa6 commented:
That is only needed if you want to be able to select as a meter on the energy panel in HA. May be useful for all Wh, kWh, MWh, VArh, kVArh, MVArh meters.
